3단계 - 질문 삭제하기 리팩터링

Jongwoo-dev opened the pull request:
안녕하세요. 마지막 단계네요.
이번엔 작업하면서 질문 두가지가 생겼습니다.

첫번째로 연관관계매핑에서 양방향 처리할때 List안에 contains 체크를 해야하나 말아야하나 고민을 하다가 넣었는데 크게 문제없는지. 그리고 다른 추가작업은 안해도 되는지가 첫번째 질문입니다. (피드백 적용 커밋이 아닌 question delete 구현 커밋에서 보셔야합니다.  앞쪽 커밋에선 실수로 느낌표를 빼먹었어서 뒤에서 수정했네요😓)

두번째로 User Entity의 equals 비교할때 id(PK 컬럼)만 비교했는데 저걸로도 괜찮을까요?
DB상으로는 pk만 같으면 같은게 맞긴한데 자바 객체에서 다른필드 다 재끼고 저것만 비교하려고 하니 뭔가 찝찝해서 여쭤봅니다.

제 질문은 여기까지 입니다.
이번에도 잘 부탁드립니다.🙏

Jongwoo-dev committed:
refactor: 피드백 적용

Jongwoo-dev committed:
doc: 이번 스탭에서 구현할 기능 목록 README.md 파일에 추가

Jongwoo-dev committed:
feat: answer delete 구현 및 단위테스트 추가

Jongwoo-dev committed:
feat: question delete 구현 및 단위테스트 추가

Jongwoo-dev committed:
feat: qnaService의 deleteQuestion 리팩토링

hhlyn0328 commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -60,4 +62,17 @@ void delete() {
     }
+
+    @Test
+    @DisplayName("질문 삭제하기 로직")
```
HappyCase 에 대한 테스트만 존재하는 것 같아요 
Question 의 유저가 다를 때, 답변 중에 작성자가 다른 사람이 1명 이상 포함될 때 실패하는 케이스가 
추가되면 좋을 것 같습니다

hhlyn0328 commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -72,4 +72,18 @@ void saveWithQuestion() {
+
+    @Test
+    @DisplayName("답변 삭제하기 로직")
+    void deleteAnswer() {
```
HappyCase 에 대한 테스트만 존재하는 것 같아요 
작성자가 다른 답변을 삭제하려고 할 때, 실패하는 케이스가 
추가되면 좋을 것 같습니다

hhlyn0328 commented on src/main/java/qna/domain/Question.java:
```diff
@@ -108,14 +122,37 @@ public void setDeleted(boolean deleted) {
+        List<DeleteHistory> deleteHistories = new ArrayList<>();
+        deleted = true;
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, id, writer, LocalDateTime.now()));
+        deleteHistories.addAll(deleteAnswers(loginUser));
```
하나의 답변 삭제라도 실패하면, 질문 삭제가 불가 하기 때문에 
답변 삭제를 먼저 시도하도록 순서를 변경하면 어떨까요?

hhlyn0328 commented on src/main/java/qna/domain/Question.java:
```diff
@@ -61,10 +72,13 @@ public Question writeBy(User writer) {
     }
 
     public void addAnswer(Answer answer) {
+        if (!this.answers.contains(answer)) {
```
OneToMany 에서 관계를 갖는 Many 데이터를 추가 할 때는 2가지 있습니다 

1. Answer에서 셋팅하는 경우 
Answer.setQuestion(question) 후, AnswerRepository.save() 합니다
이 때, 1차 캐시를 먼저 확인한 뒤에 신규 answer의 경우에는 insert 쿼리, question 변경의 경우 update 쿼리가 수행됩니다
이미 동일 question 에 매핑됐던 answer의 경우에는 1차 캐시에서 변경감지가 되지 않기 때문에 쿼리가 수행되지 않습니다 


2. Question에서 셋팅하는 경우
Question 의` List<Answer>` 속성에 cascade 설정 후, List 에 add 하여  QuestionRepository.save() 합니다 
이 때 리스트안에 동일한 answer가 있으면, 예외가 발생합니다

이처럼 중복 저장의 케이스가 발생할 수 없어서 인지, 
저는 업무를 진행하면서 contains 를 체크하고 add 했던 적은 없었던 거 같습니다.

제가 이 부분 피드백 드리려고 보다보니, 1단계에서 작성하셨어야하는 DataJpaTest 가 없네요 😮 
놓쳤었나 봅니다 ㅠ ㅜ 관련해서 DataJpaTest 를 다양하게 작성해 보신다면, 조금 더 이해가 가실 것 같아요

Jongwoo-dev replied:
답변 감사합니다~
혹시 실례가 되지 않으면 DataJpaTest 관해서 조금만 더 자세히 설명 부탁드려도 될까요?
기존 테스트에 연관관계 매핑 관련된 테스트 케이스를 더 추가했으면 하시는 말씀이신지
아니면 DataJpaTest 관련해서 제가 통째로 빼먹은 무언가가 있는건지
어떤 부분을 말씀하시는건지 잘 모르겠습니다😢

hhlyn0328 replied:
현재 실제로 DB 에 접근하는 Repository 테스트가 일체 없는데요 
1단계 미션이 JPA repository를 추가하고, 이에 대한 학습테스트로 repository 테스트를 작성해보는 것이 있었습니다 
제가 Entity 테스트만 보고 repository 테스트를 작성하셨다고 착각했어요 ㅠㅠ 
이 때, repository 테스트를 작성하려면 QuestionRepository를 @Autowired로 주입 받아야하는데 
그러려면 JpaRepository 파일마다 @repository 어노테이션을 추가해서 빈으로 정의해주셔야 하며, 
테스트 파일에도 @DataJpaTest를 명시하여, repository 빈을 주입받아 테스트를 진행함을 명시해주셔야합니다.

서비스 테스트도 현재 mocking 을 사용하여, 실제 쿼리는 수행하지 않고 있습니다 
연관관계 관련하거나, 실제로 기대했던 쿼리들이 수행되는지 보고 싶은 부분에 대해서 
테스트를 작성해볼 수 있을 것 같아요

Jongwoo-dev replied:
아! 서비스테스트도  @DataJpaTest 요거 적용하고 mocking 제거해야하는건가 보네요!
하긴 jpaTest가 가능한데 굳이 mocking으로 할 이유는 없긴하죠;;
피드백 적용하면서 같이 적용해놓겠습니다!

Jongwoo-dev replied:
으ㅠㅠ 서비스에서 mock을 날리니깐 서비스를 @InjectMocks 한 부분에서 꼬여버리네요.
서비스 테스트의 repository는 @autowired에 담기는데
injectionMock으로 들어간 서비스객체의 repository는 정상적으로 적용이 안되서요.ㅠㅠ
요부분은 좀 더 찾아봐야겠네요.ㅠㅠ

아 그리고 1단계 미션때 @DataJpaTest는 Entity 테스트에다 적용시켜버렸는데(이것때문에 헷갈리신것 같습니다) 요거 Repository 테스트는 따로 분리하는게 좋을까요??

hhlyn0328 replied:
네네 목적에 따라 분리하면 좋을 거 같아요 
도메인 테스트는 DB 접속 없이 수행되는게 맞을 거 같습니다 
repository 테스트가 추가로 잘 작성되어있다면, service 테스트에 mocking 제거는 필수는 아닙니다 😅

Jongwoo-dev replied:
아하! 그럼 repository 테스트 분리한 후 케이스 좀 더 추가하고
service 테스트는 원복하면 되겠네요
답변 감사합니다!

hhlyn0328 commented on src/main/java/qna/domain/User.java:
```diff
@@ -137,6 +137,21 @@ public void setEmail(String email) {
+        if (o == null || getClass() != o.getClass())
+            return false;
+        User user = (User)o;
+        return id.equals(user.id);
```
JPA 에서 관리되는 객체들은 key를 기반으로 1차캐시 및 변경감지를 하기 때문에 
이미 ID 로 equals 여부를 판단하고 있다고 생각하시면 될 것 같습니다!

hhlyn0328 reviewed (changes requested):
3단계 질문 삭제하기 리팩터링을 잘 진행해주셨네요 👍 

질문 주셨던 내용에 대한 답변과 
지난 리뷰에서 누락된 것 같은 피드백을 남겨보았습니다 

조금만 더 고민해보고 JPA 마무리하겠습니다
화이팅 🔥

Jongwoo-dev committed:
test: 피드백적용 - AnswerTest 에 답변삭제하기 로직 실패 케이스 테스트 추가 및 테스트 오류 수정

Jongwoo-dev committed:
test: 피드백적용 - QuestionTest 에 답변삭제하기 로직 실패 케이스 테스트 추가 및 테스트 오류 수정

Jongwoo-dev committed:
test: 피드백적용 - Question Entity의 addAnswer contains 체크 과정 제거

Jongwoo-dev committed:
test: 피드백적용 - Question Entity의 delete 메소드에서 answers의 삭제를 먼저 시도하게끔 수정

Jongwoo-dev committed:
feat: Answer Entity의 Equals와 hashcode메소드 오버라이드

Jongwoo-dev committed:
test: 피드백적용 - AnswerTest에서 model 테스트와 repository 테스트를 분리

Jongwoo-dev committed:
test: 피드백적용 - QuestionTest에서 model 테스트와 repository 테스트를 분리

Jongwoo-dev committed:
test: 피드백적용 - DeleteHistoryTest에서 model 테스트와 repository 테스트를 분리

Jongwoo-dev committed:
test: 피드백적용 - DeleteHistory Entity Test에서 사용하지않는 @DataJpaTest 제거

Jongwoo-dev committed:
test: 피드백적용 - UserTest에서 model 테스트와 repository 테스트를 분리

Jongwoo-dev committed:
test: 피드백적용 - User Entity Test에 단위 테스트 추가

Jongwoo-dev committed:
test: 지금까지의 변경점에 맞춰서 QnaServiceTest 수정

1. 연관관계 매핑 설정 구현시 자동으로 Answer가 Question에 들어가게 처리된 것을 반영하여 테스트코드에서 addAnswer 라인 제거
2. Question Entity의 delete 피드백 적용(삭제 순서 변경)에 따른 테스트 검증 데이터 변경(deleteHistories의 순서 변경)

hhlyn0328 reviewed (approved):
피드백 반영을 잘 해주셨네요 👍 

이것으로 JPA미션은 마무리하도록 하겠습니다 !
다른 미션 진행하시면서도 지난 미션에 대해 궁금한 점이 생기면 언제든 DM주세요!

과정의 끝까지 무사 완주하시길 응원하겠습니다 🔥
